deploy/ray-operator/controllers/raycluster_controller.go

```
@@ -0,0 +1,238 @@
+		// create pod if not exist
+		if !utils.IsCreated(&replica) {
+			log.Info("Creating pod", "index", i, "create pod", replica.Name)
+			if err := r.Create(context.TODO(), &replica); err != nil {
+				return reconcile.Result{}, err
+			}
+			// pod created, no more work possible for this round
+			continue
+		}
+	}
+
+	// delete pods to desired state
+	if difference.Cardinality() > 0 {
+		log.Info("difference", "pods", difference)
+		for _, runtimePod := range runtimePods.Items {
+			if difference.Contains(runtimePod.Name) {
+				log.Info("Deleting pod", "namespace", runtimePod.Namespace, "name", runtimePod.Name)
+				if err := r.Delete(context.TODO(), &runtimePod); err != nil {
+					return reconcile.Result{}, err
+				}
+				if strings.EqualFold(runtimePod.Labels[common.ClusterPodType], common.Head) {
+					svcConf := common.DefaultServiceConfig(*instance, runtimePod.Name)
+					raySvcHead := common.ServiceForPod(svcConf)
+					log.Info("delete head service", "headName", runtimePod.Name)
+					if err := r.Delete(context.TODO(), raySvcHead); err != nil {
+						return reconcile.Result{}, err
+					}
+				}
+			}
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// Build cluster instance pods.
+func (r *RayClusterReconciler) buildPods(instance *rayiov1alpha1.RayCluster) []corev1.Pod {
+	var pods []corev1.Pod
+	if instance.Spec.Extensions != nil && len(instance.Spec.Extensions) > 0 {
+		for _, extension := range instance.Spec.Extensions {
+			var i int32 = 0
+			for i = 0; i < *extension.Replicas; i++ {
+				podType := fmt.Sprintf("%v", extension.Type)
+				podName := instance.Name + common.DashSymbol + extension.GroupName + common.DashSymbol + podType + common.DashSymbol + utils.FormatInt32(i)
+				podConf := common.DefaultPodConfig(instance, podType, podName)
+				podConf.Extension = extension
+				pod := common.BuildPod(podConf)
+				pods = append(pods, *pod)
+			}
+		}
+	} else {
+		log.Info("RayCluster extensions are nil or empty")
```

Should we return an error?

---

In some cases, user will create a RayCluster without pods, and scale out in a while.

---

Then I think we should add a warning event to the object instead of outputing a info log in controller side.

---

Yes, it's better to add warning event.I will do it.